### ISSUE TITLE
Grabber doesn't handle "data:" URLs correctly

### ISSUE THREAD

**Issue by hadess** [jangernert/FeedReader#495]
It seems that this feed:
`http://www.hongkiat.com/blog/feed/`
uses quite a few embedded images, resulting in those errors:
```
Jun 01 16:14:03 classic org.gnome.FeedReader.Daemon[1623]: [ DEBUG ] GrabberUtils: download Image http://www.hongkiat.com/data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==
Jun 01 16:14:03 classic org.gnome.FeedReader.Daemon[1623]: [ ERROR ] Error downloading image: http://www.hongkiat.com/data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==
```

There are usually missing images in the posts, but I'm not sure whether those are the same ones that it embeds as "data".

**Comment by jangernert** [comment]
Please post a full example with html source code (only relevant lines), the correct url to the image and how to get it.

**Comment by bil-elmoussaoui** [comment]
@jangernert Here you go
```html
<img alt="Embedded Image" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADIA..." />
```
The image mimetype is`image/png` the base64 means that the image is encoded using base64 (must be decoded before saving as a file). And the content is `iVBORw0KGgoAAAANSUhEUgAAADIA...`
The best idea is to parse img src and replace it from data... to a full path url of the cached image

**Comment by hadess** [comment]
I think that WebKit should be able to handle URLs from this scheme without any changes if "same-origin" is used, so it would just be a matter of the grabber ignoring the "data:" scheme ones.

**Comment by bil-elmoussaoui** [comment]
@hadess yeah that will work, but FeedReader caches the images to be able to read your article offline

**Comment by hadess** [comment]
They're in the HTML, so they're already as available offline as the page itself...

**Comment by bil-elmoussaoui** [comment]
Never mind what i said, little tired :p

**Comment by jangernert** [comment]
@hadess the image data is embedded in the HTML file as base64 encoded string? I don't see where the images should be.
Also this would only work if there is ftr-site-config file to grab the relevant content. As soon as there is such a thing the config file specifies which parts of the html code to rip out to only get the relevant article text.
Are you using master or stable branch? Because in master the grabber doesn't ignore <noscript> anymore. Since the images are available in the <noscript> tag it would be fine to just ignore the data-stuff.

**Comment by hadess** [comment]
> @hadess the image data is embedded in the HTML file as base64 encoded string? I don't see where the images should be.

I think what's happening is that the grabber thinks it's a relative URL, and makes it absolute, and tries to fetch the image, which obviously doesn't work. The URL gets rewritten in the grabbed HTML, and points to a local file that doesn't exist, hence the broken images.

Does that sound plausible?

> Are you using master or stable branch?

I'm using the stable Flatpak.

> Because in master the grabber doesn't ignore anymore.

I don't understand this, doesn't ignore what?

> Since the images are available in the tag it would be fine to just ignore the data-stuff.

Right, but I think they still get rewritten. Is there an easy way to fetch the rewritten file on-disk directly?

**Comment by jangernert** [comment]
@hadess github didn't print `<noscript>` until i put it into a code block.
But still, where exactly is the image located the base64 string is pointing to?

**Comment by bil-elmoussaoui** [comment]
@jangernert It doesn't point to anywhere, on HTML & CSS you can add a png image by using an encrypted buffer with base64. So it's the image itself, if you copy the data decrypt it using base64 decode and saved the result as a png file you will get the image.
I think that the content grabber should just treat images from a src that starts with http/https..

**Comment by hadess** [comment]
`<noscript>` doesn't block "data:" images. Also see https://en.wikipedia.org/wiki/Data_URI_scheme for a longer explanation than Bilal's :)

**Comment by jangernert** [comment]
I know what was confusing me. The encoded image is just a tiny loading gif or something similar. I was wondering how a large image could be reduced to such a short string. The actual image is in `data-lazy-src`

**Comment by hadess** [comment]
> +		if(url == null || url.down().index_of('data:image') != -1)

Why index_of(...) != -1 ? Shouldn't data:image be the first thing in the URL, not something somewhere in the middle of the string?

**Comment by bil-elmoussaoui** [comment]
@hadess index_of will return -1 if it's not found. There isn't any predefined string method equivalent to `startswith` on Python or another programming language;

**Comment by bil-elmoussaoui** [comment]
@hadess i have improved that and replaced index_of with has_prefix

**Comment by hadess** [comment]
I guess `index_of(...) == 0` is the same as `startswith` or `has_prefix` (the string matches at index 0). But thanks for the fix! (in commit 8061f92a68bce39949c469a9d56e5bba799d2b26 FTR)
